x-pack/plugins/security/server/lib/authorization/has_privileges.js

```diff
@@ -0,0 +1,107 @@
+  const kibanaVersion = config.get('pkg.version');
+  const application = config.get('xpack.security.rbac.application');
+  const kibanaIndex = config.get('kibana.index');
+
+  const loginPrivilege = getLoginPrivilege();
+  const versionPrivilege = getVersionPrivilege(kibanaVersion);
+
+  return function hasPrivilegesWithRequest(request) {
+
+    const hasApplicationPrivileges = async (privileges) => {
+      const privilegeCheck = await callWithRequest(request, 'shield.hasPrivileges', {
+        body: {
+          applications: [{
+            application,
+            resources: [DEFAULT_RESOURCE],
+            privileges
+          }]
+        }
+      });
+
+      const hasPrivileges = privilegeCheck.application[application][DEFAULT_RESOURCE];
+
+      // We include the login action in all privileges, so the existence of it and not the version privilege
+      // lets us know that we're running in an incorrect configuration. Without the login privilege check, we wouldn't
+      // know whether the user just wasn't authorized for this instance of Kibana in general
+      if (!hasPrivileges[getVersionPrivilege(kibanaVersion)] && hasPrivileges[getLoginPrivilege()]) {
+        throw new Error('Multiple versions of Kibana are running against the same Elasticsearch cluster, unable to authorize user.');
+      }
+
+      return {
+        username: privilegeCheck.username,
+        hasAllRequested: privilegeCheck.has_all_requested,
+        privileges: hasPrivileges
+      };
+    };
+
+    const hasPrivilegesOnKibanaIndex = async () => {
+      const privilegeCheck = await callWithRequest(request, 'shield.hasPrivileges', {
+        body: {
+          index: [{
+            names: [kibanaIndex],
+            privileges: ['create', 'delete', 'read', 'view_index_metadata']
+          }]
+        }
+      });
+
+      return Object.values(privilegeCheck.index[kibanaIndex]).includes(true);
+    };
+
+    return async function hasPrivileges(privileges) {
+      const allPrivileges = [versionPrivilege, loginPrivilege, ...privileges];
+      const privilegesCheck = await hasApplicationPrivileges(allPrivileges);
+
+      const username = privilegesCheck.username;
+
+      // We don't want to expose the version privilege to consumers, as it's an implementation detail only to detect version mismatch
+      const missing = Object.keys(privilegesCheck.privileges)
+        .filter(p => !privilegesCheck.privileges[p])
+        .filter(p => p !== versionPrivilege);
+
+      if (privilegesCheck.hasAllRequested) {
+        return {
+          result: HAS_PRIVILEGES_RESULT.AUTHORIZED,
+          username,
+          missing,
+        };
+      }
+
+      if (!privilegesCheck.privileges[loginPrivilege] && await hasPrivilegesOnKibanaIndex()) {
+        const msg = `Relying on implicit privileges determined from the index privileges is deprecated and will be removed in Kibana 7.0`;
+        server.log(['warning', 'deprecated', 'security'], msg);
```

Comment:
I worry about this becoming unbelievably chatty. If you had a few hundred users and triggered this legacy behavior, your log file would be pretty unusable. Any thoughts on how we could make this less chatty?

Reply:
Not really because logging these deprecations only on login isn't really feasible, since this behavior could change during a user's "session". We could potentially only log these deprecations a maximum amount of time per user. It doesn't feel like the chattyness of the current solution justifies a complex solution when they can easily get away from the deprecation warnings by using the Kibana Application Privileges.

Reply:
> Not really because logging these deprecations only on login isn't really feasible, since this behavior could change during a user's "session".

But that would only happen if a person for some reason went backwards from the new model to the old. This doesn't seem likely to me, and even in the event that it did happen, they'd still get the log on the subsequent login for that user and any other user's login that has that role combination.

The deprecation logs are suppose to flag to an administrator that something should be updated on their install in a non-emergency capacity. If there are rare edge cases that might mean that they don't get as many deprecation notices as they would otherwise, that still seems preferable to me to having an upgrade of Kibana trigger a massive flood of deprecation notices.

> It doesn't feel like the chattyness of the current solution justifies a complex solution when they can easily get away from the deprecation warnings by using the Kibana Application Privileges.

I think you're understating the effort involved in migrating to this new model for large deployments. In addition to learning about the new model, vigorous verification and testing of the new model before rolling it out, admins will need to come up with a strategy for updating potentially dozens/hundreds/thousands(?) of roles across their organization. It's not like we're providing an automatic mechanism for them to switch all their existing roles over to this new system.